README.md
# memtest
Utility for testing RAM on your mac system. Supports OS X 10.7 and above. This is a contents of dmg image obtained from http://www.memtestosx.org

Site above requires a small fee to be paid in order to download, but this software is licensed under GPLv2 and provided here for free.

## Usage
Just run `./memtest`

In order to test as much memory as possible, it is advised to run memtest in single user mode:

1. To boot into single user mode, reboot you mac and hold `command-S`
2. Mount filesystem `/sbin/mount -uw /`
3. Go to directory where you cloned the repo `cd /Users/user/this_repo_path/memtest`
4. Run memtest `./memtest all 3 -l` - this command will run test 3 times, using all available memory and saving output to log file


Charles Cazabon's Info/CHANGELOG
Version 4.04M  (Macintosh port of memtester 4.0.4)
05 September 2004 Tony Scaminaci

  - Changed memory reduction algorithm for single-user mode
    from a fixed 17M to a percentage (2.3%) of installed memory
    in order to eliminate lock-ups of the Darwin kernel

  - made memtester's exit code meaningful.  See the manpage for its meaning.
    Thanks to Wurzel Parsons-Keir, who sent a patch for the code, so I only had
    to document it (Charles).


Version 4.03M (Macintosh port of memtester 4.0.3)
28 August 2004 Tony Scaminaci

  - ported version 4.0.3 to MacOS X and added functionality
    to enable operation in single-user mode

Version 4.0.3
10 August 2004

  -small changes to enable building with dietlibc and a few other environments
  that don't even attempt to provide the various Posix definitions.
  -cosmetic fixes to output.
  -restore the reduce-and-retry loop of memory locking from version 2.

Version 4.0.2
9 August 2004

  -add manpage

Version 4.0.1
8 August 2004

  -fix cosmetic bugs in output

Version 4.0.0
7 August 2004

  -rewrite to clean up the code (previously an ugly hack), for 64-bit
  cleanliness
  -change build system to build on (hopefully) any platform.  Previous
  versions required hackery on some systems.

Version 3 not publicly released.


source/memtest.h
/*
 * Very simple but very effective user-space memory tester.
 * Originally by Simon Kirby <sim@stormix.com> <sim@neato.org>
 * Version 2 by Charles Cazabon <memtest@discworld.dyndns.org>
 * Version 3 not publicly released.
 * Version 4 rewrite:
 * Copyright (C) 2004 Charles Cazabon <memtest@discworld.dyndns.org>
 * Copyright (C) 2004 - 2014 Tony Scaminaci (Macintosh ports)
 * Licensed under the terms of the GNU General Public License version 2 (only).
 * See the file COPYING for details.
 */

// Macro definitions for handling 32/64 bit platforms at compile time

#include <limits.h>

#define rand32() ((unsigned int) rand() | ( (unsigned int) rand() << 16))

#if (ULONG_MAX == 4294967295UL)
    #define rand_ul() rand32()
    #define UL_ONEBITS 0xffffffff
    #define UL_LEN 32
    #define CHECKERBOARD1 0x55555555
    #define CHECKERBOARD2 0xaaaaaaaa
    #define UL_BYTE(x) ((x | x << 8 | x << 16 | x << 24))
#elif (ULONG_MAX == 18446744073709551615ULL)
    #define rand64() (((ul) rand32()) << 32 | ((ul) rand32()))
    #define rand_ul() rand64()
    #define UL_ONEBITS 0xffffffffffffffffUL
    #define UL_LEN 64
    #define CHECKERBOARD1 0x5555555555555555
    #define CHECKERBOARD2 0xaaaaaaaaaaaaaaaa
    #define UL_BYTE(x) (((ul)x | (ul)x<<8 | (ul)x<<16 | (ul)x<<24 | (ul)x<<32 | (ul)x<<40 | (ul)x<<48 | (ul)x<<56))
#else
    #error long on this platform is not 32 or 64 bits
#endif

// Mask bit definitions for command line arguments
// Tony Scaminaci (updated 1/2007)

#define USE_LOG_FILE		0x00000001
#define QUICK_FLAG			0x00000002
#define RESERVED_2			0X00000004
#define RESERVED_3			0X00000008
#define NUMBER_OF_PASSES	0x00000FF0
#define	TEST_SIZE_MB		0xFFFFF000

// Mask bits used in Linear PRN test
// Tony Scaminaci 10/2006

#define Tap63Mask	0x4000000000000000
#define Tap1Mask	0x0000000000000001

// Custom type definitions

typedef unsigned long ul;
typedef unsigned long long ull;
typedef unsigned long volatile ulv;

struct test
{
	char *name;
	int (*fp)();
};

// Test routine declarations

int test_stuck_address(unsigned long volatile *buf, int arg_list, size_t count);					// Linear stuck address test - full buffer
int test_linear_prn(unsigned long volatile *buf, int arg_list, size_t count, time_t timestamp);		// Linear pseudorandom test - full buffer
int test_random_value(unsigned long volatile *bufa, unsigned long volatile *bufb, int arg_list, size_t count);
int test_xor_comparison(unsigned long volatile *bufa, unsigned long volatile *bufb, int arg_list, size_t count);
int test_sub_comparison(unsigned long volatile *bufa, unsigned long volatile *bufb, int arg_list, size_t count);
int test_mul_comparison(unsigned long volatile *bufa, unsigned long volatile *bufb, int arg_list, size_t count);
int test_div_comparison(unsigned long volatile *bufa, unsigned long volatile *bufb, int arg_list, size_t count);
int test_or_comparison(unsigned long volatile *bufa, unsigned long volatile *bufb, int arg_list, size_t count);
int test_and_comparison(unsigned long volatile *bufa, unsigned long volatile *bufb, int arg_list, size_t count);
int test_seqinc_comparison(unsigned long volatile *bufa, unsigned long volatile *bufb, int arg_list, size_t count);
int test_solidbits_comparison(unsigned long volatile *bufa, unsigned long volatile *bufb, int arg_list, size_t count);
int test_checkerboard_comparison(unsigned long volatile *bufa, unsigned long volatile *bufb, int arg_list, size_t count);
int test_blockseq_comparison(unsigned long volatile *bufa, unsigned long volatile *bufb, int arg_list, size_t count);
int test_walkbits0_comparison(unsigned long volatile *bufa, unsigned long volatile *bufb, int arg_list, size_t count);
int test_walkbits1_comparison(unsigned long volatile *bufa, unsigned long volatile *bufb, int arg_list, size_t count);
int test_bitspread_comparison(unsigned long volatile *bufa, unsigned long volatile *bufb, int arg_list, size_t count);
int test_bitflip_comparison(unsigned long volatile *bufa, unsigned long volatile *bufb, int arg_list, size_t count);



source/tests_original.c
/*
 * Very simple but very effective user-space memory tester.
 * Originally by Simon Kirby <sim@stormix.com> <sim@neato.org>
 * Version 2 by Charles Cazabon <memtest@discworld.dyndns.org>
 * Version 3 not publicly released.
 * Version 4 rewrite:
 * Copyright (C) 2004 Charles Cazabon <memtest@discworld.dyndns.org>
 * Portions Copyright (C) 2005 Tony Scaminaci
 * Licensed under the terms of the GNU General Public License version 2 (only).
 * See the file COPYING for details.
 *
 * This file contains the functions for the actual tests, called from the
 * main routine in memtest.c.  See other comments in that file.
 *
 */

#include <sys/types.h>
#include <stdio.h>
#include <stdlib.h>

#include "memtest.h"

// Global variables

extern FILE	*logfile;

char progress[] = "-\\|/";
#define PROGRESSLEN 4
#define PROGRESSOFTEN 600000

/* Function definitions. */

int compare_regions(ulv *bufa, ulv *bufb, size_t count)
{
    int r = 0;
    size_t i;

    for (i = 0; i < count; i++)
	{
        if (bufa[i] != bufb[i])
		{
            fprintf(stderr, "FAILURE: 0x%08lx != 0x%08lx at offset 0x%08lx.\n", (ul) bufa[i], (ul) bufb[i], (ul) i);
			if (logfile != NULL)
			  {
				fprintf(logfile, "FAILURE: 0x%08lx != 0x%08lx at offset 0x%08lx.\n", (ul) bufa[i], (ul) bufb[i], (ul) i);
				fflush(logfile);
			  }
            r = -1;
        }
    }
    return r;
}

int test_stuck_address(ulv *buf, size_t count)
{
    unsigned int j;
    size_t i;

	printf("           ");
    fflush(stdout);
    for (j = 0; j < 16; j++)
	{
	    printf("\b\b\b\b\b\b\b\b\b\b\b");
        printf("setting %3u", j);
        fflush(stdout);
        for (i = 0; i < count; i++)
		{
            buf[i] = ((j + i) % 2) == 0 ? (ul) &buf[i] : ~((ul) &buf[i]);
        }
        printf("\b\b\b\b\b\b\b\b\b\b\b");
        printf("testing %3u", j);
        fflush(stdout);
        for (i = 0; i < count; i++)
		{
            if (buf[i] != (((j + i) % 2) == 0 ? (ul) &buf[i] : ~((ul) &buf[i])))
			{
                fprintf(stderr, "FAILURE: possible bad address line at offset 0x%08lx.\n", (ul) i);
                printf("Skipping to next test...\n");
				fflush(stdout);
				if (logfile != NULL)
				  {
					fprintf(logfile, "FAILURE: possible bad address line at offset 0x%08lx.\n", (ul) i);
					fprintf(logfile, "Skipping to next test...\n");
					fflush(logfile);
				  }
                return -1;
            }
        }
    }
    printf("\b\b\b\b\b\b\b\b\b\b\b           \b\b\b\b\b\b\b\b\b\b\b");
    fflush(stdout);
    return 0;
}

int test_random_value(ulv *bufa, ulv *bufb, size_t count)
{
    ul j = 0;
    size_t i;

	putchar(' ');
    fflush(stdout);
    for (i = 0; i < count; i++)
	{
        bufa[i] = bufb[i] = rand_ul();
		if (!(i % PROGRESSOFTEN))
		{
			putchar('\b');
			putchar(progress[++j % PROGRESSLEN]);
            fflush(stdout);
		}
    }
	printf("\b \b");
    fflush(stdout);
    return compare_regions(bufa, bufb, count);
}

int test_xor_comparison(ulv *bufa, ulv *bufb, size_t count)
{
    size_t i;
    ul q = rand_ul();

    for (i = 0; i < count; i++)
	{
        bufa[i] ^= q;
        bufb[i] ^= q;
    }
    return compare_regions(bufa, bufb, count);
}

int test_sub_comparison(ulv *bufa, ulv *bufb, size_t count)
{
    size_t i;
    ul q = rand_ul();

    for (i = 0; i < count; i++)
	{
        bufa[i] -= q;
        bufb[i] -= q;
    }
    return compare_regions(bufa, bufb, count);
}

int test_mul_comparison(ulv *bufa, ulv *bufb, size_t count)
{
    size_t i;
    ul q = rand_ul();

    for (i = 0; i < count; i++)
	{
        bufa[i] *= q;
        bufb[i] *= q;
    }
    return compare_regions(bufa, bufb, count);
}

int test_div_comparison(ulv *bufa, ulv *bufb, size_t count)
{
    size_t i;
    ul q = rand_ul();

    for (i = 0; i < count; i++)
	{
        if (!q)
		{
            q++;
        }
        bufa[i] /= q;
        bufb[i] /= q;
    }
    return compare_regions(bufa, bufb, count);
}

int test_or_comparison(ulv *bufa, ulv *bufb, size_t count)
{
    size_t i;
    ul q = rand_ul();

    for (i = 0; i < count; i++)
	{
        bufa[i] |= q;
        bufb[i] |= q;
    }
    return compare_regions(bufa, bufb, count);
}

int test_and_comparison(ulv *bufa, ulv *bufb, size_t count)
{
    size_t i;
    ul q = rand_ul();

    for (i = 0; i < count; i++)
	{
        bufa[i] &= q;
        bufb[i] &= q;
    }
    return compare_regions(bufa, bufb, count);
}

int test_seqinc_comparison(ulv *bufa, ulv *bufb, size_t count)
{
    size_t i;
    ul q = rand_ul();

    for (i = 0; i < count; i++)
	{
        bufa[i] = bufb[i] = (i + q);
    }
    return compare_regions(bufa, bufb, count);
}

int test_solidbits_comparison(ulv *bufa, ulv *bufb, size_t count)
{
    unsigned int j;
    ul q;
    size_t i;

	printf("           ");
    fflush(stdout);
    for (j = 0; j < 64; j++)
	{
	    printf("\b\b\b\b\b\b\b\b\b\b\b");
        q = (j % 2) == 0 ? UL_ONEBITS : 0;
        printf("setting %3u", j);
        fflush(stdout);
        for (i = 0; i < count; i++)
		{
            bufa[i] = bufb[i] = (i % 2) == 0 ? q : ~q;
        }
        printf("\b\b\b\b\b\b\b\b\b\b\b");
        printf("testing %3u", j);
        fflush(stdout);
        if (compare_regions(bufa, bufb, count))
		{
            return -1;
        }
    }
    printf("\b\b\b\b\b\b\b\b\b\b\b           \b\b\b\b\b\b\b\b\b\b\b");
    fflush(stdout);
    return 0;
}

int test_checkerboard_comparison(ulv *bufa, ulv *bufb, size_t count)
{
    unsigned int j;
    ul q;
    size_t i;

	printf("           ");
    fflush(stdout);
    for (j = 0; j < 64; j++)
	{
	    printf("\b\b\b\b\b\b\b\b\b\b\b");
        q = (j % 2) == 0 ? CHECKERBOARD1 : CHECKERBOARD2;
        printf("setting %3u", j);
        fflush(stdout);
        for (i = 0; i < count; i++)
		{
            bufa[i] = bufb[i] = (i % 2) == 0 ? q : ~q;
        }
        printf("\b\b\b\b\b\b\b\b\b\b\b");
        printf("testing %3u", j);
        fflush(stdout);
        if (compare_regions(bufa, bufb, count))
		{
            return -1;
        }
    }
    printf("\b\b\b\b\b\b\b\b\b\b\b           \b\b\b\b\b\b\b\b\b\b\b");
    fflush(stdout);
    return 0;
}

int test_blockseq_comparison(ulv *bufa, ulv *bufb, size_t count)
{
    unsigned int j;
    size_t i;

	printf("           ");
    fflush(stdout);
    for (j = 0; j < 256; j++)
	{
	    printf("\b\b\b\b\b\b\b\b\b\b\b");
        printf("setting %3u", j);
        fflush(stdout);
        for (i = 0; i < count; i++)
		{
            bufa[i] = bufb[i] = (ul) UL_BYTE(j);
        }
        printf("\b\b\b\b\b\b\b\b\b\b\b");
        printf("testing %3u", j);
        fflush(stdout);
        if (compare_regions(bufa, bufb, count))
		{
            return -1;
        }
    }
    printf("\b\b\b\b\b\b\b\b\b\b\b           \b\b\b\b\b\b\b\b\b\b\b");
    fflush(stdout);
    return 0;
}

int test_walkbits0_comparison(ulv *bufa, ulv *bufb, size_t count)
{
    unsigned int j;
    size_t i;

	printf("           ");
    fflush(stdout);
    for (j = 0; j < 64; j++)
	{
	    printf("\b\b\b\b\b\b\b\b\b\b\b");
        printf("setting %3u", j);
        fflush(stdout);
        for (i = 0; i < count; i++)
		{
            if (j < UL_LEN) 
			{ /* Walk it up. */
                bufa[i] = bufb[i] = 0x00000001 << j;
            } else
			{ /* Walk it back down. */
                bufa[i] = bufb[i] = 0x00000001 << (UL_LEN * 2 - j - 1);
            }
        }
        printf("\b\b\b\b\b\b\b\b\b\b\b");
        printf("testing %3u", j);
        fflush(stdout);
        if (compare_regions(bufa, bufb, count))
		{
            return -1;
        }
    }
    printf("\b\b\b\b\b\b\b\b\b\b\b           \b\b\b\b\b\b\b\b\b\b\b");
    fflush(stdout);
    return 0;
}

int test_walkbits1_comparison(ulv *bufa, ulv *bufb, size_t count)
{
    unsigned int j;
    size_t i;

	printf("           ");
    fflush(stdout);
    for (j = 0; j < UL_LEN * 2; j++)
	{
	    printf("\b\b\b\b\b\b\b\b\b\b\b");
		printf("setting %3u", j);
        fflush(stdout);
        for (i = 0; i < count; i++)
		{
            if (j < UL_LEN)
			{ /* Walk it up. */
                bufa[i] = bufb[i] = UL_ONEBITS ^ (0x00000001 << j);
            } else
			{ /* Walk it back down. */
                bufa[i] = bufb[i] = UL_ONEBITS ^ (0x00000001 << (UL_LEN * 2 - j - 1));
            }
        }
        printf("\b\b\b\b\b\b\b\b\b\b\b");
        printf("testing %3u", j);
        fflush(stdout);
        if (compare_regions(bufa, bufb, count))
		{
            return -1;
        }
    }
    printf("\b\b\b\b\b\b\b\b\b\b\b           \b\b\b\b\b\b\b\b\b\b\b");
    fflush(stdout);
    return 0;
}

int test_bitspread_comparison(ulv *bufa, ulv *bufb, size_t count)
{
    unsigned int j;
    size_t i;

	printf("           ");
    fflush(stdout);
    for (j = 0; j < UL_LEN * 2; j++)
	{
	    printf("\b\b\b\b\b\b\b\b\b\b\b");
        printf("setting %3u", j);
        fflush(stdout);
        for (i = 0; i < count; i++)
		{
            if (j < UL_LEN)
			{ /* Walk it up. */
                bufa[i] = bufb[i] = (i % 2 == 0)
                    ? (0x00000001 << j) | (0x00000001 << (j + 2))
                    : UL_ONEBITS ^ ((0x00000001 << j)
                                    | (0x00000001 << (j + 2)));
            } else
			{ /* Walk it back down. */
                bufa[i] = bufb[i] = (i % 2 == 0)
                    ? (0x00000001 << (UL_LEN * 2 - 1 - j)) | (0x00000001 << (UL_LEN * 2 + 1 - j))
                    : UL_ONEBITS ^ (0x00000001 << (UL_LEN * 2 - 1 - j)
                                    | (0x00000001 << (UL_LEN * 2 + 1 - j)));
            }
        }
        printf("\b\b\b\b\b\b\b\b\b\b\b");
        printf("testing %3u", j);
        fflush(stdout);
        if (compare_regions(bufa, bufb, count))
		{
            return -1;
        }
    }
    printf("\b\b\b\b\b\b\b\b\b\b\b           \b\b\b\b\b\b\b\b\b\b\b");
    fflush(stdout);
    return 0;
}

int test_bitflip_comparison(ulv *bufa, ulv *bufb, size_t count)
{
    unsigned int j, k;
    ul q;
    size_t i;

	printf("           ");
    fflush(stdout);
    for (k = 0; k < UL_LEN; k++)
	{
        q = 0x00000001 << k;
        for (j = 0; j < 8; j++)
		{
    	    printf("\b\b\b\b\b\b\b\b\b\b\b");
            q = ~q;
            printf("setting %3u", k * 8 + j);
            fflush(stdout);
            for (i = 0; i < count; i++) {
                bufa[i] = bufb[i] = (i % 2) == 0 ? q : ~q;
            }
            printf("\b\b\b\b\b\b\b\b\b\b\b");
            printf("testing %3u", k * 8 + j);
            fflush(stdout);
            if (compare_regions(bufa, bufb, count))
			{
                return -1;
            }
        }
    }
    printf("\b\b\b\b\b\b\b\b\b\b\b           \b\b\b\b\b\b\b\b\b\b\b");
    fflush(stdout);
    return 0;
}
